scripts/benchmark.py

```diff
@@ -0,0 +1,238 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+import argparse
+import json
+import logging
+import platform
+import sys
+import time
+import traceback
+from datetime import datetime
+from pathlib import Path
+from typing import Callable, List
+
+import czifile
+import numpy as np
+import psutil
+from quilt3 import Package
+from tqdm import tqdm
+
+import aicspylibczi as aics
+
+###############################################################################
+
+logging.basicConfig(
+    level=logging.INFO,
+    format="[%(levelname)4s: %(module)s:%(lineno)4s %(asctime)s] %(message)s",
+)
+log = logging.getLogger(__name__)
+
+###############################################################################
+
+CLUSTER_CONFIGS = [
+    {"name": "small-local-cluster-replica", "per_worker_cores": 2, "workers": 4},
+    {"name": "large-local-cluster-replica", "per_worker_cores": 2, "workers": 16},
+    {"name": "small-worker-distributed-cluster", "per_worker_cores": 1, "workers": 32},
+    {
+        "name": "many-small-worker-distributed-cluster",
+        "per_worker_cores": 1,
+        "workers": 128,
+    },
+    {
+        "name": "many-standard-worker-distributed-cluster",
+        "per_worker_cores": 4,
+        "workers": 32,
+    },
+]
+
+###############################################################################
+
+# Args
+
+
+class Args(argparse.Namespace):
+    def __init__(self):
+        self.__parse()
+
+    def __parse(self):
+        # Setup parser
+        p = argparse.ArgumentParser(
+            prog="benchmark_aics",
+            description=(
+                "Run read time benchmarks for aicspylibczi against other common image "
+                "readers. The benchmark dataset can be downloaded using the "
+                "download_test_resources script with the specific hash: "
+                "5e665ed66c1b373a84002227044c7a12a2ecc506b84a730442a5ed798428e26a"
+            ),
+        )
+
+        # Arguments
+        p.add_argument(
+            "--save_path",
+            default="benchmark/results.json",
+            type=Path,
+            help="Path to save the generated benchmark JSON file.",
+        )
+        p.add_argument(
+            "--upload",
+            action="store_true",
+            help="Should the results be uploaded to Quilt.",
+        )
+        p.add_argument(
+            "--debug",
+            action="store_true",
+            help="Show traceback if the script were to fail.",
+        )
+
+        # Parse
+        p.parse_args(namespace=self)
+
+
+###############################################################################
+
+
+def _run_benchmark(
+        resources_dir: Path,
+        extensions: List[str],
+        non_aics_reader: List[Callable],
+        iterations: int = 3,
+):
+    # Collect files matching the extensions provided
+    files = []
+    for ext in extensions:
+        files += list(resources_dir.glob(ext))
+        files += list(Path("/Users/jamies/Data").glob(ext))
```

Comment:
Can this be removed or perhaps hoisted out of the main implementation into a command line arg?

Reply:
No one else will be able to run this.  Just make sure that intent is clear.